### ISSUE TITLE
[Help]: Disabled zwave-js node

### ISSUE THREAD

**Issue by fabnavigator** [zwave-js/node-red-contrib-zwave-js#184]
### Contact Details (optional)

wcmarinelli@gmail.com

### How can we help?

I installed this node using npm into my node-red directory like all of my other nodes. Only three of the four nodes are enabled.

```
npm install node-red-contrib-zwave-js@6.5.5
npm rebuild
```

![zwave](https://user-images.githubusercontent.com/16625904/155454873-7213f5bc-b646-4dce-b244-96cef6c46f58.png)

### Any more details?

I run node-red in a docker container.  I also run containers for zwavejsmqtt and home-assistant. I'm hoping that I can do away with zwavejs2mqtt and home-assistant, and do everything in node-red (with the help of your nodes). The only reason I use home-assistant is to talk to zwavejs2mqtt and eventually my zwave devices. It would be great to cut out those two middle-men.

Note: nodejs and npm are not runnning in docker containers.

See below for the docker-compose code. I think I might need to add the usb device to the node-red container setup at some point.

### Version

6.5.5

### Node-RED Version

2.2.1

### What hardware are you using?

Raspberry Pi

### Any code to add?

```shell
node-red log:

Welcome to Node-RED
===================
23 Feb 22:43:29 - [info] Node-RED version: v2.2.1
23 Feb 22:43:29 - [info] Node.js  version: v12.22.8
23 Feb 22:43:29 - [info] Linux 5.10.92-v8+ arm64 LE
23 Feb 22:43:29 - [info] Loading palette nodes
23 Feb 22:43:33 - [info] Dashboard version 3.1.5 started at /ui
23 Feb 22:43:33 - [warn] ------------------------------------------------------
23 Feb 22:43:33 - [warn] [node-red-contrib-zwave-js/zwave-js] Error: The module '/data/node_modules/@serialport/bindings/build/Release/bindings.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 93. This version of Node.js requires
NODE_MODULE_VERSION 72. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
23 Feb 22:43:33 - [warn] ------------------------------------------------------
23 Feb 22:43:33 - [info] Settings file  : /data/settings.js
23 Feb 22:43:33 - [info] Context store  : 'default' [module=localfilesystem]
23 Feb 22:43:33 - [info] Context store  : 'memory' [module=memory]
23 Feb 22:43:33 - [info] User directory : /data
23 Feb 22:43:33 - [warn] Projects disabled : editorTheme.projects.enabled=false
23 Feb 22:43:33 - [info] Flows file     : /data/flows.json
23 Feb 22:43:33 - [info] Server now running at http://127.0.0.1:1880/
23 Feb 22:43:33 - [info] Starting flows
23 Feb 22:43:34 - [info] [cronplus:9:00a 25th] createTask - index: 0, static: true, opt: {"name":"schedule1","topic":"topic1","payloadType":"default","payload":"","expressionType":"cron","expression":"0 0 9 25 * * *","location":"","offset":"0","solarType":"all","solarEvents":"sunrise,sunset"}
23 Feb 22:43:34 - [info] Started flows
23 Feb 22:43:35 - [info] [server:Home Assistant] Connecting to http://home-assistant:8123
23 Feb 22:43:35 - [info] [server:Home Assistant] Connected to http://home-assistant:8123

docker-compose containers:

  zwavejs2mqtt:
    container_name: zwavejs2mqtt
    image: zwavejs/zwavejs2mqtt:6.5.2
    tty: true
    ports:
      - 8091:8091
      - 3000:3000
    volumes:
      - /home/fabnav/opt/zwavejs2mqtt:/usr/src/app/store
    devices:
      # - /dev/ttyACM0:/dev/ttyACM0 # 500 series
      - /dev/ttyUSB0:/dev/ttyUSB0 # 700 series
      
  home-assistant:
    container_name: home-assistant
    image: homeassistant/raspberrypi4-64-homeassistant:2022.2.9
    privileged: true
    depends_on:
      - zwavejs2mqtt
    ports:
      - 8123:8123
    volumes:
      - /home/fabnav/opt/home-assistant:/config
      - /etc/localtime:/etc/localtime:ro

  node-red:
    container_name: node-red
    image: nodered/node-red:2.2.1-12
    depends_on:
      - postgres
      - home-assistant
    user: '1001'
    ports:
      - 1880:1880
    volumes:
      - /home/fabnav/opt/node-red:/data
      - /etc/localtime:/etc/localtime:ro
    environment:
      TZ: "America/New_York"
```

**Comment by marcus-j-davies** [comment]
Hi @fabnavigator.

So...
I know nothing about HA or Docker, but I think I can see what is happening.

Node RED is using Node JS 12.22.8
`
23 Feb 22:43:29 - [info] Node.js  version: v12.22.8
`  - So far so good.

Our Node RED zwave js module uses the `serialport` package and this package employs native NodeJS modules which are compiled against a certain NodeJS version during install time.

Now here is the problem:
Node RED is using `v12.22.8` - (NODE_MODULE_VERSION 72)
but the `serialport `package was compiled against >= `v16.0.0` - (NODE_MODULE_VERSION 93)

```
23 Feb 22:43:33 - [warn] ------------------------------------------------------
23 Feb 22:43:33 - [warn] [node-red-contrib-zwave-js/zwave-js] Error: The module '/data/node_modules/@serialport/bindings/build/Release/bindings.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 93. This version of Node.js requires                           
NODE_MODULE_VERSION 72. Please try re-compiling or re-installing                             
the module (for instance, using `npm rebuild` or `npm install`).
23 Feb 22:43:33 - [warn] ------------------------------------------------------
```

Therefore, it looks as if the install happened under a different NodeJS version that Node RED is actually running under.
did you do the install, 'outside' of the container - is that right word ?? 🤷‍♂️ 

Do this, 'outside' the node RED container `node -v ` and see what version you get.

The error log relating to different versions that was used during build and what is actually running - is certainly the problem.
the node that is disabled - is the the one using `serialport`

**Comment by fabnavigator** [comment]
Thank you. Indeed the nodejs version inside of the container is 12.22.8 and outside is 16.14.0. Node-red is of course using the inside version.

Option A - Learn more about docker so that I can rebuild the image with the newer version of nodejs.
Option B - Run node-red outside of docker (ugh).

Before I spend time on this, do you think I will be able to fully control my zwave devices using these nodes in node-red? I need to trigger nodes when device states change, and of course fully control them, including zwave device configuration.

**Comment by fabnavigator** [comment]
See above. THank you again.

**Comment by marcus-j-davies** [comment]
Hi @fabnavigator 

> do you think I will be able to fully control my zwave devices using these nodes in node-red

Yes.
The module has a User Interface, that takes care of setting configuration options for your devices  (via the Configuration CC), but it can also be used to control the command classes in general (see marked areas below)

The user interface is the control centre for your zwave network, from adding, removing, healing and controlling devices.
make sure to provide any security/encryption keys on the controller node.

There is also examples that come with this node - to get your started 👍 

![image](https://user-images.githubusercontent.com/55892693/155577535-c890b864-e993-4307-ba72-ecf43291a734.png)

>  I need to trigger nodes when device states change, and of course fully control them

Yes.
ALL network events are injected into your flow from the main controller node, and device-nodes.

Many many events can be consumed within Node RED
This wiki page has all the possible events, and details how they get to you, as well as the message format to use to control devices in a flow (i.e not using the UI)

https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/Payload-Messaging-Format

> Option A - Learn more about docker so that I can rebuild the image with the newer version of nodejs.
> Option B - Run node-red outside of docker (ugh).

You might have an Option C 😄 

Use the `exec` node and get it to run the following command.
`npm rebuild serialport` - this may actually build the serial module using the Node RED container environment. - meaning it use the same node version as Node RED 😉

**Comment by fabnavigator** [comment]
Yes. I guess we were typing at the same time.

**Comment by marcus-j-davies** [comment]
Feel free to spark up a conversation if you need, until then, happy flow building.

**Comment by fabnavigator** [comment]
I'm not up and running quite yet. zwave-js can't access my seial port. Should I open a new issue, or just continue here?
I gave the serial port permissions that should allow the container to use it.
```crw-rw-rw- 1 root dialout 188, 0 Feb 17 13:50 /dev/ttyUSB0```
With my other zwave container I had to add:
```
    devices:
     - /dev/ttyUSB0:/dev/ttyUSB0
```
to my docker -compose file. I tried it for the node-red container and it isn't working. node-red is crashing.
